lib/cache/cache.go

```diff
@@ -1294,43 +1285,73 @@ func (c *Cache) GetNodes(ctx context.Context, namespace string, opts ...services
 	return rg.presence.GetNodes(ctx, namespace, opts...)
 }
 
+// getNodesWithFnCache implements TTL-based caching for the GetNodes endpoint.  All nodes that will be returned from the caching layer
+// must be cloned to avoid concurrent modification.
+func (c *Cache) getNodesWithFnCache(ctx context.Context, rg readGuard, namespace string, opts ...services.MarshalOption) ([]types.Server, error) {
```

Comment:
nit: personally i feel a little unsafe passing a copy of readGuard around. it's ok as long we release outside, not inside.

Reply:
Yeah, its always released by the function that originally acquired it.  I agree that its a bit funky, but I also take some comfort in the fact that `readGuard` has a really specific usage pattern that all methods follow identically, so I think the chances of accidental misuse here are very low.